[dynamo] handle min/max on arbitrary iterables

bdhirsh opened the pull request:
Allows us to not graph break on code like:
```
max_val = max(a.size(0) for a in args)
```

This showed up in `speech_transformer` at least (probably a few other models too).

There's an API mismatch between python and FX, where python's min/max can take in arbitrary lists to reduce over, but our SymInt.max() only handles pairs of ints. I added a bridge in dynamo that handles an arbitrary list, and reduces over the elements.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #92643
* #92618
* #92617



cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

bdhirsh committed:
[dynamo] handle min/max on arbitrary iterables

[ghstack-poisoned]

bdhirsh commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -358,7 +358,39 @@ def call_function(
+            else:
+                return variables.ConstantVariable(min(a.value, b.value))
+
+        # torch min/max only support two arguments, but python min/max support a variable number.
```
Every time I make a change in dynamo, I question whether or not the added complexity is because I'm doing it wrong, or of something fundamental in the dynamo -> torch translation.

This one feels more fundamental, but pls tell me if otherwise (cc @voznesenskym)

jansel commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -358,7 +358,39 @@ def call_function(
 
-    def _call_min_max(self, tx, a, b):
+    def _call_min_max(self, tx, *args):
+        if all(isinstance(a, variables.ConstantVariable) for a in args):
```
Is this case actually needed?  I'd expect just the two arg version to work fine with the old code.

jansel commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -358,7 +358,39 @@ def call_function(
+
+        # torch min/max only support two arguments, but python min/max support a variable number.
+        # When given more than 2, reduce over the arguments.
+        if len(args) != 2:
+            if len(args) == 1 and isinstance(args[0], variables.BaseListVariable):
+                args_to_reduce = args[0].items
+            elif len(args) == 1 and isinstance(args[0], variables.ListIteratorVariable):
+                # Need to treat ListIterator specially from BaseList: max() on a python iterator will "use up" the iterator.
+                # Is there / should there be a more idiomatic way to do this?
+                # Maybe manually evaluating an interator in dynamo is uncommon.
+                args_to_reduce = []
+                curr_iter = args[0]
+                while curr_iter.index < len(curr_iter.items):
+                    val, next_iter = curr_iter.next_variables()
+                    args_to_reduce.append(val)
+                    tx.replace_all(curr_iter, next_iter)
+                    curr_iter = next_iter
+            else:
+                assert len(args) > 2, f"min/max was given an invalid set of arguments: {', '.join([type(x) for x in args])}"
+                args_to_reduce = args
+            # manually handle the base case for lists: max(single_element_list) = single_element_list[0]
+            if len(args_to_reduce) == 1:
+                return args_to_reduce[0]
+            return functools.reduce(lambda a, b: self._call_min_max(tx, a, b), args_to_reduce)
```
I think this can be simplified, something like:

```suggestion
        if len(args) == 1:
            return self._call_min_max(tx, *args[0].unpack_var_sequence(tx))
        elif len(args) > 2:
            a, b, *rest = args
            return self._call_min_max(
                tx,
                self._call_min_max(tx, a, b),
                *rest)
```

bdhirsh committed:
Update on "[dynamo] handle min/max on arbitrary iterables"

Allows us to not graph break on code like:
```
max_val = max(a.size(0) for a in args)
```

This showed up in `speech_transformer` at least (probably a few other models too).

There's an API mismatch between python and FX, where python's min/max can take in arbitrary lists to reduce over, but our SymInt.max() only handles pairs of ints. I added a bridge in dynamo that handles an arbitrary list, and reduces over the elements.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "[dynamo] handle min/max on arbitrary iterables"

Allows us to not graph break on code like:
```
max_val = max(a.size(0) for a in args)
```

This showed up in `speech_transformer` at least (probably a few other models too).

There's an API mismatch between python and FX, where python's min/max can take in arbitrary lists to reduce over, but our SymInt.max() only handles pairs of ints. I added a bridge in dynamo that handles an arbitrary list, and reduces over the elements.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] handle min/max on arbitrary iterables"

Allows us to not graph break on code like:
```
max_val = max(a.size(0) for a in args)
```

This showed up in `speech_transformer` at least (probably a few other models too).

There's an API mismatch between python and FX, where python's min/max can take in arbitrary lists to reduce over, but our SymInt.max() only handles pairs of ints. I added a bridge in dynamo that handles an arbitrary list, and reduces over the elements.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] handle min/max on arbitrary iterables"

Allows us to not graph break on code like:
```
max_val = max(a.size(0) for a in args)
```

This showed up in `speech_transformer` at least (probably a few other models too).

There's an API mismatch between python and FX, where python's min/max can take in arbitrary lists to reduce over, but our SymInt.max() only handles pairs of ints. I added a bridge in dynamo that handles an arbitrary list, and reduces over the elements.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] handle min/max on arbitrary iterables"

Allows us to not graph break on code like:
```
max_val = max(a.size(0) for a in args)
```

This showed up in `speech_transformer` at least (probably a few other models too).

There's an API mismatch between python and FX, where python's min/max can take in arbitrary lists to reduce over, but our SymInt.max() only handles pairs of ints. I added a bridge in dynamo that handles an arbitrary list, and reduces over the elements.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] handle min/max on arbitrary iterables"

Allows us to not graph break on code like:
```
max_val = max(a.size(0) for a in args)
```

This showed up in `speech_transformer` at least (probably a few other models too).

There's an API mismatch between python and FX, where python's min/max can take in arbitrary lists to reduce over, but our SymInt.max() only handles pairs of ints. I added a bridge in dynamo that handles an arbitrary list, and reduces over the elements.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] handle min/max on arbitrary iterables"

Allows us to not graph break on code like:
```
max_val = max(a.size(0) for a in args)
```

This showed up in `speech_transformer` at least (probably a few other models too).

There's an API mismatch between python and FX, where python's min/max can take in arbitrary lists to reduce over, but our SymInt.max() only handles pairs of ints. I added a bridge in dynamo that handles an arbitrary list, and reduces over the elements.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] handle min/max on arbitrary iterables"

Allows us to not graph break on code like:
```
max_val = max(a.size(0) for a in args)
```

This showed up in `speech_transformer` at least (probably a few other models too).

There's an API mismatch between python and FX, where python's min/max can take in arbitrary lists to reduce over, but our SymInt.max() only handles pairs of ints. I added a bridge in dynamo that handles an arbitrary list, and reduces over the elements.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] handle min/max on arbitrary iterables"

Allows us to not graph break on code like:
```
max_val = max(a.size(0) for a in args)
```

This showed up in `speech_transformer` at least (probably a few other models too).

There's an API mismatch between python and FX, where python's min/max can take in arbitrary lists to reduce over, but our SymInt.max() only handles pairs of ints. I added a bridge in dynamo that handles an arbitrary list, and reduces over the elements.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] handle min/max on arbitrary iterables"

Allows us to not graph break on code like:
```
max_val = max(a.size(0) for a in args)
```

This showed up in `speech_transformer` at least (probably a few other models too).

There's an API mismatch between python and FX, where python's min/max can take in arbitrary lists to reduce over, but our SymInt.max() only handles pairs of ints. I added a bridge in dynamo that handles an arbitrary list, and reduces over the elements.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] handle min/max on arbitrary iterables"

Allows us to not graph break on code like:
```
max_val = max(a.size(0) for a in args)
```

This showed up in `speech_transformer` at least (probably a few other models too).

There's an API mismatch between python and FX, where python's min/max can take in arbitrary lists to reduce over, but our SymInt.max() only handles pairs of ints. I added a bridge in dynamo that handles an arbitrary list, and reduces over the elements.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] handle min/max on arbitrary iterables"

Allows us to not graph break on code like:
```
max_val = max(a.size(0) for a in args)
```

This showed up in `speech_transformer` at least (probably a few other models too).

There's an API mismatch between python and FX, where python's min/max can take in arbitrary lists to reduce over, but our SymInt.max() only handles pairs of ints. I added a bridge in dynamo that handles an arbitrary list, and reduces over the elements.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] handle min/max on arbitrary iterables"

Allows us to not graph break on code like:
```
max_val = max(a.size(0) for a in args)
```

This showed up in `speech_transformer` at least (probably a few other models too).

There's an API mismatch between python and FX, where python's min/max can take in arbitrary lists to reduce over, but our SymInt.max() only handles pairs of ints. I added a bridge in dynamo that handles an arbitrary list, and reduces over the elements.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] handle min/max on arbitrary iterables"

Allows us to not graph break on code like:
```
max_val = max(a.size(0) for a in args)
```

This showed up in `speech_transformer` at least (probably a few other models too).

There's an API mismatch between python and FX, where python's min/max can take in arbitrary lists to reduce over, but our SymInt.max() only handles pairs of ints. I added a bridge in dynamo that handles an arbitrary list, and reduces over the elements.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] handle min/max on arbitrary iterables"

Allows us to not graph break on code like:
```
max_val = max(a.size(0) for a in args)
```

This showed up in `speech_transformer` at least (probably a few other models too).

There's an API mismatch between python and FX, where python's min/max can take in arbitrary lists to reduce over, but our SymInt.max() only handles pairs of ints. I added a bridge in dynamo that handles an arbitrary list, and reduces over the elements.




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]
